proposals/001-file-format-api.md

```diff
@@ -0,0 +1,608 @@
+
+
+## Motivation
+
+* The way we handle formats is currently not unified across platforms.
+* We don't have a documented strategy to determine the format of a file, which might cause differences across implementations.
+* We are only using the media type and file extensions to determine the format, which is not enough in some cases, e.g. ZIP archives without extension.
+* It's currently not easy for a reading app to support a custom publication format in Readium.
+* Formats supported by Readium are currently exposed as enums with different cases on each platform, which is confusing.
+
+```swift
+// Swift
+enum Publication.Format {
+    /// Formats natively supported by Readium.
+    case cbz, epub, pdf, webpub
+    /// Default value when the format is not specified.
+    case unknown
+}
+```
+
+```kotlin
+// Kotlin
+enum class Publication.TYPE {
+    EPUB, CBZ, FXL, WEBPUB, AUDIO, DiViNa
+}
+```
+
+## Proposed Solution
+
+If a given profile of a file format has its own media type, it means that it is worth differenciating for reading apps. Therefore, **we declare one `Format` per media type**, instead of:
+
+* grouping several similar formats under the same `Format` – e.g. webpub, audiobook and DiViNa manifests, or
+* splitting a unique format into two sub-formats – e.g. EPUB reflowable and fixed-layout.
+
+This proposal adds the following new types:
+
+* [`MediaType`](#mediatype-class) offers parsing and comparison of media types.
+* [`Format`](#format-class) represents a known file format, uniquely identified by a media type.
+* [`Format.Sniffer`](#formatsniffer-interface) is a function type used to implement the sniffing algorithm of each format.
+* [`File`](#file-class) offers a convenient way to cache a format used by several Readium components.
+
+
+### `MediaType` class
+
+Comparing media types is more complicated than it looks, [since they can contain parameters](https://tools.ietf.org/html/rfc6838) such as `charset=utf-8`. We can't ignore them because some formats use parameters in their media type, for example `application/atom+xml;profile=opds-catalog` for an OPDS 1 catalog.
+
+#### Constructor
+
+* `MediaType(string: String)`
+  * Creates a `MediaType` from its string representation.
+
+#### Properties
+
+* `string: String` (or `toString()` if more idiomatic).
+  * The string representation of this media type.
+* `type: String`
+  * `application` in `application/epub+zip`
+* `subtype: String`
+  * `epub+zip` in `application/epub+zip`
+* `parameters: Map<String, String>`
+  * The parameters in the media type, such as `charset=utf-8`.
+
+#### Helpers
+
+More can be added as needed.
+
+* `isManifest: Bool`
+  * Returns whether this media type is contained by `AudiobookManifest`, `DiViNaManifest`, `Opds1Entry`, `Opds2Publication` or `WebPubManifest`.
+  * Used to determine if the remote URL needs to be stored in the database, instead of the local downloaded path.
+* `isOpds: Bool`
```

Comment:
Maybe a more generic name could be used: `Catalog` for example. OPDS is just one format.

Reply:
Actually the use case is that a reading app will want to present a catalog with different views, depending on whether it's an HTML page or an OPDS catalog. So we need a way to differentiate them.

Hence, the "used to determine the type of remote catalogs" for both `isHtml` and `isOpds`.

Reply:
Hum, ok.